### PR TITLE
Add example to evaluate sky shader on CPU 

### PR DESCRIPTION
**Pull request by repi**
Adds a new example that shows how to compile and evaluate a simple fragment shader (our sky shader) directly on the CPU.

For this specific shader as it doesn't use any resources it is very easy as one can just call the fragment shader function manually, later for standard shaders with resource bindings this will be much harder, but maybe still useful to test parts of shaders or reuse some functionality.

Right now it also just works for the current temp math functions that use `core` intrinsic, as all of those are available obviously on the CPU also, later when we start introducing GPU & SPIRV specific intrinsics (#81), this approach won't work, and then we will want to use a higher-level math layer in the shaders for the math instructions (glam: #134) so they can be mapped to the right available target intrinsic.

To make the example very simple, slim and cross-platform, this example uses the `minifb` crate to drive the Window. Could switch to something else later though but it is very basic.

image

### CODE REVIEW

File: examples/example-shader/src/lib.rs [hunk 1]
```diff
@@ -168,7 +190,15 @@ pub fn main_fs(input: Input<Vec4>, mut output: Output<Vec4>) {
...
 
-    output.store(color.extend(0.0))
+    color.extend(1.0)
+}
+
+#[allow(unused_attributes)]
+#[spirv(entry = "fragment")]
+pub fn main_fs(input: Input<Vec4>, mut output: Output<Vec4>) {
+    let v = input.load();
+    let color = fs(Vec2::new(v.0, v.1));
```
Comment (khyperia) [inline examples/example-shader/src/lib.rs:200]: This makes me think that we should talk about the unwrapping of Inputs/Outputs more - I've flipflopped back and forth on this, there's arguments both ways. Also, we should be able to construct Intputs/Outputs on the CPU, right? (maybe need to make something `pub`, `pub fn construct_from_cpu_var(&mut T) -> Output<T>` or something)

File: examples/example-runner-cpu/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    window.limit_update_rate(Some(std::time::Duration::from_micros(16600)));
+
+    let start_time = Instant::now();
+
+    let buffer = (0..WIDTH * HEIGHT)
+        .into_par_iter()
+        .map(|i| {
+            let screen_pos = vec2(
+                (i % WIDTH) as f32 / WIDTH as f32 * 2.0 - 1.0,
+                (i / HEIGHT) as f32 / HEIGHT as f32 * 2.0 - 1.0,
```
Comment (khyperia) [inline examples/example-runner-cpu/src/main.rs:34]: I... thought this should be `(i / WIDTH)`, and I'm confused how this works at all (is the image shifted slightly compared to the GPU version?), but maybe it's just me just waking up.

**Review by khyperia — APPROVED** [review]
this is *super* cool, and impressively simple!

> later when we start introducing GPU & SPIRV specific intrinsics (#81), this approach won't work

It *should* work, I plan on implementing it in a way where it should work.

File: examples/example-runner-cpu/src/main.rs (hunk 2 above)
Comment (repi) [reply examples/example-runner-cpu/src/main.rs:34]: Just you waking up I think :) this is correct and works. Can likely be made cleaner in a separate functions or so maybe later, but does work

File: examples/example-shader/src/lib.rs (hunk 1 above)
Comment (repi) [reply examples/example-shader/src/lib.rs:200]: Yes the current `Input` and `Output` was not ideal for this, but do think it is good to have this use case at least and we can think about the design separately

File: examples/example-runner-cpu/src/main.rs (hunk 2 above)
Comment (bjorn3) [reply examples/example-runner-cpu/src/main.rs:34]: It took me a while to wrap my head around this, but what happens is the following: The `i % WIDTH` generates an x coordinate in the range `0..WIDTH`, the `/ WIDTH` turns it into the range `0.0 .. 1.0`, the `* 2.0 - 1.0` then turns it into the range `-1.0 .. 1.0`. Same for the y coordinate.

Comment (khyperia) [reply examples/example-runner-cpu/src/main.rs:34]: okay no I'm more awake now and it's *not* just me waking up, this *should* be `(i / WIDTH) as f32 / HEIGHT as f32 * 2.0 - 1.0`.

This PR's CPU code:

cpu_original

GPU code:

gpu

This PR's CPU code but with the line changed to `i / WIDTH`:

cpu_changed

it is indeed "the image shifted slightly compared to the GPU version"

Comment (repi) [reply examples/example-runner-cpu/src/main.rs:34]: Hah awesome, I was overconfident that my code worked, well it did work just not correctly. Thanks!!

**Review by Jasper-Bekkers — APPROVED** [review]
I love that we can do this :-)

**Comment by repi** [comment]
Ah some cargo-deny stuff to fix with the minifb dependency

**Comment by repi** [comment]
@Mergifyio rebase

File: examples/example-runner-cpu/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
+use minifb::{Key, Window, WindowOptions};
+use rayon::prelude::*;
+use spirv_std::{vec2, Vec4};
+use std::time::Instant;
+
+fn color_u32_from_vec4(v: Vec4) -> u32 {
```
Comment (emilk) [inline examples/example-runner-cpu/src/main.rs:6]: Maybe a comment here about why it is OK to ignore sRGB here (or add a TODO about respecting sRGB) ?

Comment (repi) [reply examples/example-runner-cpu/src/main.rs:6]: done

**Pull request merged**